lightning/src/util/persist.rs

```diff
@@ -194,3 +206,857 @@ where
+/// and sorting names of monitors and updates.
+mod types {
+	use crate::chain::transaction::OutPoint;
+	use crate::util::persist::KVStoreUpdatingPersisterError;
+	use bitcoin::hashes::hex::{FromHex, ToHex};
+	use bitcoin::Txid;
+	use core::convert::TryFrom;
+	use std::io::Error;
+
+	/// A struct representing a name for a monitor.
+	#[derive(Clone, Debug)]
+	pub struct MonitorName(String);
+
+	impl MonitorName {
+		/// Verifies that an [`OutPoint`] can be formed from the given `name`.
+		pub fn new(name: String) -> Result<Self, Error> {
+			let me = Self(name);
+			me.do_try_into_outpoint()?;
+			Ok(me)
+		}
+		/// Convert this monitor name to a str.
+		pub fn as_str(&self) -> &str {
+			&self.0
+		}
+		fn do_try_into_outpoint(&self) -> Result<OutPoint, Error> {
+			let mut parts = self.0.splitn(2, '_');
+			let txid_hex =
+				parts
+					.next()
+					.ok_or_else(|| KVStoreUpdatingPersisterError::BadMonitorName {
+						reason: "name is not a splittable string",
+						context: &self.0,
+					})?;
+			let index =
+				parts
+					.next()
+					.ok_or_else(|| KVStoreUpdatingPersisterError::BadMonitorName {
+						reason: "no index value found after underscore",
+						context: &self.0,
+					})?;
+			let index =
+				index
+					.parse()
+					.map_err(|_| KVStoreUpdatingPersisterError::BadMonitorName {
+						reason: "could not parse index value in monitor name",
+						context: &self.0,
+					})?;
+			let txid = Txid::from_hex(txid_hex).map_err(|_| {
+				KVStoreUpdatingPersisterError::BadMonitorName {
+					reason: "bad txid in monitor name",
+					context: &self.0,
+				}
+			})?;
+			Ok(OutPoint { txid, index })
+		}
+	}
+
+	impl TryFrom<MonitorName> for OutPoint {
+		type Error = Error;
+
+		fn try_from(value: MonitorName) -> Result<Self, Error> {
+			value.do_try_into_outpoint()
+		}
+	}
+
+	impl From<OutPoint> for MonitorName {
+		fn from(value: OutPoint) -> Self {
+			MonitorName(format!("{}_{}", value.txid.to_hex(), value.index))
+		}
+	}
+
+	/// A struct representing a name for an update.
+	#[derive(Clone, Debug)]
+	pub struct UpdateName(u64, String);
+
+	impl UpdateName {
+		/// Validates that an update sequence ID can be derived from the given `name`.
+		pub fn new(name: String) -> Result<Self, Error> {
+			match name.parse::<u64>() {
+				Ok(u) => Ok(Self(u, UpdateName::left_padded_string(u))),
+				Err(_) => Err(KVStoreUpdatingPersisterError::BadUpdateName {
+					reason: "cannot parse u64 from update name",
+					context: &name,
+				}
+				.into()),
+			}
+		}
+
+		/// Convert this monitor update name to a &str
+		pub fn as_str(&self) -> &str {
+			&self.1
+		}
+
+		/// Left-pad the sequential update id for string representation
+		fn left_padded_string(value: u64) -> String {
+			format!("{:0>20}", value)
+		}
+	}
+
+	impl From<u64> for UpdateName {
+		fn from(value: u64) -> Self {
+			Self(value, UpdateName::left_padded_string(value))
+		}
+	}
+
+	impl PartialEq for UpdateName {
```

Comment:
They probably can -- I'm being paranoid here, mostly out of learning-Rustness. I ended up making this a `(u64, String)`, as the `as_str` method can't form a `String` and then return the reference to the temporary value, so I needed the `UpdateName` to own the `String`. At that point, I was concerned about corner cases with a derived sort of `(u64, String)`. On reflection and further reading, I think I can trust the derived sort here because it will only look at the string _if_ you somehow got two update names that parsed as the same `u64`. To make a long story short, I probably don't need to worry about that (one's storage is truly scarfed if that happens), and am adding nothing if the `KVStore.list()` is lexicographically sorted now, anyway.